include/sampleflow/consumers/average_cosinus.h

```diff
@@ -0,0 +1,205 @@
+ * cosinus length can done by setting it in mcmc_test.cc.
+ *
+ * Notice, that for each new sample $x_n$, we need to take a sample, that was l earlier than new sample ($x_{n-l}).
+ * Working this way, updating algorithm becomes very similar to mean_value.h algorithm.
+ *
+ * Everytime, for updation we calculate corresponding fractions. There numerator is dot product, while denominator -
+ * multiplied those two vector norms
+ *
+ * ### Threading model ###
+ *
+ * The implementation of this class is thread-safe, i.e., its
+ * consume() member function can be called concurrently and from multiple
+ * threads.
+   **/
+
+    template <typename InputType>
+    class AverageCosinus: public Consumer<InputType>
+    {
+      public:
+        /**
+         * The data type of the elements of the input type.
+         */
+
+       using scalar_type = typename InputType::value_type;
+
+       /**
+   	 * The data type, where we save multiple set of past samples.
+   	 */
+
+       using value_type = boost::numeric::ublas::matrix<scalar_type>;
+
+        /**
+         * Constructor.
+         */
+
+        AverageCosinus(unsigned int length);
+
+        /**
+         * Process one sample by updating the previously computed covariance
+         * matrix using this one sample.
+         *
+         * @param[in] sample The sample to process.
+         * @param[in] aux_data Auxiliary data about this sample. The current
+         *   class does not know what to do with any such data and consequently
+         *   simply ignores it.
+         */
+        virtual
+        void
+        consume (InputType     sample,
+                 AuxiliaryData aux_data) override;
+
+        /**
+         * A function that returns the average cosinus vector computed from the
+         * samples seen so far. If no samples have been processed so far, then
+         * a default-constructed object of type InputType will be returned.
+         *
+         * @return The computed average cosinus vector.
+         */
+        std::vector<scalar_type> get() const;
+
+      private:
+        /**
+         * A mutex used to lock access to all member variables when running
+         * on multiple threads.
+         */
+
+      mutable std::mutex mutex;
+
+
+        /**
+         * Parts for running average_cosinus
+         * Description of these parts is given above
+         */
+
+      std::vector<scalar_type> current_avg_cosinus;//Current average cosinus
+
+        /**
+        * Save previous sample value
+        * %%%%%%%%%%%%Manty's comment: I didn't find (I'm 100% there are) better way to save last k sample values
+        * as just using one more object and assign i'th row of past_sample to i+1'th row of past_sample_replace
+        *
+        */
+
+      value_type past_sample;
+      value_type past_sample_replace;
+
+        /**
+         * The number of samples processed so far.
+         */
+
+        types::sample_index n_samples;
+
+        unsigned int avg_cosinus_length;
```

Comment:
I think you want to call this `history_length` or some such. I was confused by what the length here refers to.